File: scripts/oop-practice-sources/swift/smartphone.swift
class Smartphone {
    private let model: String
    private var battery: Int

    init(model: String) {
        self.model = model
        self.battery = 20
    }

    func call() {
        battery = max(0, battery - 5)
        print("Звонок с \(model)... Заряд: \(battery)%")
    }

    func charge() {
        battery = min(100, battery + 30)
        print("Зарядка \(model)... Заряд: \(battery)%")
    }

    func showStatus() {
        print("Смартфон \(model): заряд \(battery)%")
    }
}

let phone = Smartphone(model: "Pixel")
phone.showStatus()
phone.call()
phone.charge()
phone.showStatus()

File: scripts/oop-practice-sources/rust/smartphone.rs
struct Smartphone {
    model: String,
    battery: i32,
}

impl Smartphone {
    fn new(model: &str) -> Self {
        Self {
            model: model.to_string(),
            battery: 20,
        }
    }

    fn call(&mut self) {
        self.battery = (self.battery - 5).max(0);
        println!("Звонок с {}... Заряд: {}%", self.model, self.battery);
    }

    fn charge(&mut self) {
        self.battery = (self.battery + 30).min(100);
        println!("Зарядка {}... Заряд: {}%", self.model, self.battery);
    }

    fn show_status(&self) {
        println!("Смартфон {}: заряд {}%", self.model, self.battery);
    }
}

fn main() {
    let mut phone = Smartphone::new("Pixel");
    phone.show_status();
    phone.call();
    phone.charge();
    phone.show_status();
}
